nexus/src/app/sagas/vpc_create.rs

```diff
@@ -333,3 +407,323 @@ async fn svc_notify_sleds(
+        assert!(no_subnets_exist(datastore).await);
+        assert!(no_firewall_rules_exist(datastore).await);
+    }
+
+    async fn no_vpcs_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::Vpc;
+        use crate::db::schema::vpc::dsl;
+
+        dsl::vpc
+            .filter(dsl::time_deleted.is_null())
+            .select(Vpc::as_select())
+            .first_async::<Vpc>(datastore.pool_for_tests().await.unwrap())
+            .await
+            .optional()
+            .unwrap()
+            .map(|vpc| {
+                eprintln!("VPC exists: {vpc:?}");
+            })
+            .is_none()
+    }
+
+    async fn no_routers_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::VpcRouter;
+        use crate::db::schema::vpc_router::dsl;
+
+        dsl::vpc_router
+            .filter(dsl::time_deleted.is_null())
+            .select(VpcRouter::as_select())
+            .first_async::<VpcRouter>(datastore.pool_for_tests().await.unwrap())
+            .await
+            .optional()
+            .unwrap()
+            .map(|router| {
+                eprintln!("Router exists: {router:?}");
+            })
+            .is_none()
+    }
+
+    async fn no_routes_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::RouterRoute;
+        use crate::db::schema::router_route::dsl;
+
+        dsl::router_route
+            .filter(dsl::time_deleted.is_null())
+            .select(RouterRoute::as_select())
+            .first_async::<RouterRoute>(
+                datastore.pool_for_tests().await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .map(|route| {
+                eprintln!("Route exists: {route:?}");
+            })
+            .is_none()
+    }
+
+    async fn no_subnets_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::VpcSubnet;
+        use crate::db::schema::vpc_subnet::dsl;
+
+        dsl::vpc_subnet
+            .filter(dsl::time_deleted.is_null())
+            .select(VpcSubnet::as_select())
+            .first_async::<VpcSubnet>(datastore.pool_for_tests().await.unwrap())
+            .await
+            .optional()
+            .unwrap()
+            .map(|subnet| {
+                eprintln!("Subnet exists: {subnet:?}");
+            })
+            .is_none()
+    }
+
+    async fn no_firewall_rules_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::VpcFirewallRule;
+        use crate::db::schema::vpc_firewall_rule::dsl;
+
+        dsl::vpc_firewall_rule
+            .filter(dsl::time_deleted.is_null())
+            .select(VpcFirewallRule::as_select())
+            .first_async::<VpcFirewallRule>(
+                datastore.pool_for_tests().await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .map(|rule| {
+                eprintln!("Firewall rule exists: {rule:?}");
+            })
+            .is_none()
+    }
```

Comment:
FYI, if I *don't* remove the firewall rules, this fails -- unwinding the saga would still leave old firewall rules around.